Pagination

VadimKoziachiy opened the pull request:
- added pagination for article in UserController and ArticleController.

Commit:
Pagination
- added pagination for article in UserController and ArticleController.

alxanko commented on src/main/java/news/blog/com/controller/ArticleController.java:
```diff
@@ -20,9 +23,10 @@
     @GetMapping
-    public ResponseEntity<Collection<ArticleDto>> getArticles()
+    public ResponseEntity<Collection<ArticleDto>> getArticles(
+            @PageableDefault(sort = {"id"}, direction = Sort.Direction.DESC) Pageable pageable)
```
what are this annotation params?

Commit:
Tags
- tag handler added;
- you can get posts by tag;
- can get all the posts.

Commit:
Comments
- added CommentEntity;
- added CommentController;
- added CommentService and CommentRepository;
- added migrations for CommentEntity.

alxanko commented on src/main/java/news/blog/com/controller/ArticleController.java:
```diff
@@ -37,4 +42,16 @@ public ResponseEntity saveArticle(@RequestBody ArticleDto articleDto)
+    }
+
+    @GetMapping("/tags/{tag}")
+    public ResponseEntity<Collection<ArticleDto>> getArticleByTag(@PathVariable String tag)
```
it's better to use `tag` param as `RequestParam`

alxanko commented on src/main/java/news/blog/com/repository/CommentRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface CommentRepository extends CrudRepository<CommentEntity, Long>
+{
+    List<CommentEntity> findByArticle_Id(Long id);
```
this method can be declared w/o underscore

alxanko commented on src/main/java/news/blog/com/service/impl/ArticleServiceImpl.java:
```diff
@@ -24,16 +27,44 @@
+                                   .collect(Collectors.toSet());
+        tags.forEach(tag ->
+                {
+                   Long quantity = articles.stream()
```
it's a complex operation
let's get unique tags directly from repositrory
and then we will count tags again with repository

alxanko commented on src/main/resources/migrations/article/init-table.xml:
```diff
@@ -11,7 +11,7 @@
                 <constraints primaryKey="true" nullable="false" unique="true"/>
             </column>
-            <column name="tags" type="varchar"/>
+            <column name="tag" type="varchar"/>
```
stop changing existing migrations!
bad practice

VadimKoziachiy replied:
will be done:)

Commit:
Correction of defects.

alxanko commented on src/main/java/news/blog/com/controller/ArticleController.java:
```diff
@@ -37,8 +37,8 @@ public ResponseEntity saveArticle(@RequestBody ArticleDto articleDto)
 
-    @GetMapping("/{id}")
-    public ResponseEntity<ArticleDto> getArticle(@PathVariable Long id)
+    @GetMapping("/article")
```
you can unite `getArticle` & `getArticleByTag` controller methods

alxanko commented on src/main/java/news/blog/com/controller/CommentController.java:
```diff
@@ -14,14 +14,14 @@
 
-    @GetMapping("/{id}")
-    public ResponseEntity getCommentsByArticle(@PathVariable Long id)
+    @GetMapping
```
it is not proper to do this

alxanko commented on src/main/java/news/blog/com/controller/CommentController.java:
```diff
@@ -14,14 +14,14 @@
 
-    @PostMapping("/{id}")
-    public ResponseEntity saveCommentsByArticle(@RequestBody CommentDto commentDto, @PathVariable Long id)
+    @PostMapping
```
it is not proper to do this

alxanko commented on src/main/java/news/blog/com/repository/ArticleRepository.java:
```diff
@@ -4,17 +4,26 @@
     List<ArticleEntity> findByTag(String string);
 
+    @Query(value = "SELECT tag FROM ArticleEntity ORDER BY tag ASC")
+    Set<String> findAllTags();
```
let's read about `distinct`

Commit:
changed sql request.
